### PR TITLE
protoc: parser rejects explicit use of map_entry option

### PR DESCRIPTION
**Pull request by jhump**
This addresses #13441.

This preserves the similar check at the _point of use_ of invalid messages in `DescriptorBuilder` (and there's an existing test that verifies that check still works).

But it adds another check in the parser, to catch this error at the _point of definition_ of an invalid message. And the corresponding test is updated: we no longer need a usage of the message to catch the error, and the reported position is the definition of the option, not the usage site of the message.

The way this works feels kinda gross, but I wasn't sure of a better way to do it. The only place we know for certain that it was an explicit option (vs. auto-added by the parser when synthesizing a map entry message) is when after processing the message body, we can look at the uninterpreted options. So that's what this does. If you have ideas on better/cleaner approaches, I'd be happy to revise.

### CODE REVIEW

File: src/google/protobuf/compiler/parser.cc [hunk 1]
```diff
@@ -931,6 +931,26 @@ bool Parser::ParseMessageBlock(DescriptorProto* message,
   if (message->reserved_range_size() > 0) {
     AdjustReservedRangesWithMaxEndNumber(message);
   }
+
+  // check if map_entry option was explicitly set
```
Comment (fowles) [inline src/google/protobuf/compiler/parser.cc:935]: rather than a comment to section off part of the message, please use a well named helper

Comment (jhump) [reply src/google/protobuf/compiler/parser.cc]: Done. I've added a `ValidateMessage` helper, analogous to existing `ValidateEnum`. The only check it performs is the check for an explicit `map_entry` option.

WDYT? Too generic?

**Review by fowles — APPROVED** [review]

File: src/google/protobuf/compiler/parser.cc (hunk 1 above)
Comment (fowles) [reply src/google/protobuf/compiler/parser.cc]: nah, looks about right.

**Comment by jhump** [comment]
@fowles, so does that mean that there are places in the google3 codebase that were using explicit `map_entry` options? And the other failing build means reserved ranges where start and end were reversed?

BTW, sorry if this makes more work for your team. These are obviously not urgent fixes. And thanks for your help in shepherding these fixes into the codebase!

**Comment by fowles** [comment]
yes, the cleanups are simple so I should be able to slice them in this week

**Pull request closed without merging**